Add `concretise` and `containerise` aliases for our UK users

tgamblin opened the pull request:
From a user:

> Aargh.
> ```
> ==> Error: concretise is not a recognized Spack command or extension command; check with `spack commands`.
> ```

To make things easier for our friends in the UK, this adds `concretise` and `containerise` aliases for the `spack concretize` and `spack containerize` commands.

tgamblin committed:
Add `concretise` and `containerise` aliases for our UK users

From a user:

> Aargh.
> ```
> ==> Error: concretise is not a recognized Spack command or extension command; check with `spack commands`.
> ```

To make things easier for our friends in the UK, this adds `concretise` and
`containerise` aliases for the `spack concretize` and `spack containerize` commands.

- [x] add aliases
- [x] update completions

haampie reviewed (dismissed):
Can it be skipped in tab completion?

```
$ spack conc<tab>
$ spack concreti<tab><tab>
concretise  concretize  
```
:weary:

tgamblin commented:
Oh that’s annoying. I think so. We generate the tab completion so we can add some attributes to the alias table to ignore these.

tgamblin commented:
How about: “if the only completions are a command and it’s always aliases, ignore the aliases”?

So `concretis<tab>` would still work. Probably not an issue *here* but I think the logic generalizes better than excepting some aliases (as sometimes you may want them to complete).

tgamblin commented:
Of course sadly thinking about it more I dunno if I can actually implement that. I’ll do exceptions if not.

haampie commented:
Alternatively there's https://github.com/spack/spack/pull/17229 so it can be user option

sethrj commented:
Is there a python "localization" (localisation?) library that could help? Check if the time zone offset is zero or one? 😂

tgamblin commented:
I don't think this is a localization issue b/c it's about a command that is basically part of Spack's public API.  You don't want that to change by locale.

adamjstewart reviewed (dismissed):
I vote we merge this and solve tab completion another day. I would prefer to keep tab completion as simple as possible as it's already complex enough that it took me a week to wrap my head around and I was the one who wrote most of it. We can't easily skip aliases because we need to be able to properly complete subcommands of those aliases.

tgamblin committed:
completion: add alias handling

Bash completion is now smarter about handling aliases. In particular, if all completions
for some input command are aliased to the same thing, we'll just complete with that thing.

If you've already *typed* the full alias for a command, we'll complete the alias.

So, for example, here there's more than one real command involved, so all aliases are
shown:

```console
$ spack con
concretise    concretize    config        containerise  containerize
```

Here, there are two possibilities: `concretise` and `concretize`, but both map to
`concretize` so we just complete that:

```console
$ spack conc
concretize
```

And here, the user has already typed `concretis`, so we just go with it as there is only
one option:

```console
 spack concretis
concretise
```

tgamblin commented:
Too late I already fixed it for `bash`.  I've implemented the following:

> Bash completion is now smarter about handling aliases. In particular, if all completions
> for some input command are aliased to the same thing, we'll just complete with that thing.
> 
> If you've already *typed* the full alias for a command, we'll complete the alias.
>
> So, for example, here there's more than one real command involved, so all aliases are shown:
> 
> ```console
> $ spack con
> concretise    concretize    config        containerise  containerize
> ```
> 
> Here, there are two possibilities: `concretise` and `concretize`, but both map to `concretize` so we just complete that:
> 
> ```console
> $ spack conc
> concretize
> ```
> 
> And here, the user has already typed `concretis`, so we just go with it as there is only one option:
> 
> ```console
>  spack concretis
> concretise
>```

tgamblin commented:
See what you think of this @adamjstewart.  I haven't updated things in the `fish` completion, but I suppose I could try. I'm ok with leaving that for a future PR.

adamjstewart commented:
I have a loose grasp on how fish completion works, it's much more complex. Volunteering this weekend but could take a look on Monday.

adamjstewart commented:
IMO the code to avoid tab-completing multiple aliases (thus saving users a single keystroke) is too complex to be worth it. So I stand by my original comment. My vote would be to undo the changes to bash. Feel free to merge as is if you want to keep it though. Just trying to keep the codebase simple and maintainable.

tgamblin commented:
I’m for keeping it. We haven’t had to touch bash completion much (at all?) since you wrote it — I think this will be similar. Also I’m not going anywhere. I can maintain that code.

tgamblin commented:
Does fish not support something similar? Or did you have time to look at it?

adamjstewart commented:
In theory, bash and fish completion are Turing complete and support anything you want, it's just a matter of how straightforward it is to do. I'm not even sure what one would search for to find the feature you're describing, all searches including the word "alias" result in the fish command by the same name.

One distinguishing factor between fish completion and bash completion is that if <kbd>Tab</kbd> outputs multiple suggestions, subsequent <kbd>Tab</kbd> will cycle through the options, so while it still requires an extra keystroke to get what you want, your finger is already there. This can be configured in bash:
```bash
bind TAB:menu-complete
```
but isn't enabled by default like it is in fish.

tgamblin committed:
completion: make bash completion work in zsh

`compgen -W` does not behave the same way in zsh as it does in bash; it seems not to
actually generate the completions we want.

- [x] add a zsh equivalent and `_compgen_w` to abstract it away
- [x] use `_compgen_w` instead of `compgen -W`

tgamblin commented:
I think I've addressed @haampie's concerns, and I've added shell tests for the new stuff.  I didn't do fish (yet) but I'm sort of inclined to learn it b/c I played around with it and it was pretty awesome.  I'll save that for another PR.

I also added some code to make bash completion work in both zsh and bash.  It previously only sort-of worked in `zsh` -- in particular `compgen -W` doesn't work in `zsh` even with `bashcompinit` -- at least not on my mac.  So I put in some code to make it work in `zsh`. I may come back to do proper `zsh` completions too -- they are not dissimilar to `fish` completions.

This at least fixes @haampie's issue for `bash` and `zsh` (which I think is important if only because `concretize` is a very common command), and `fish` users have the "hit tab twice" workaround by default so I don't think it's as important there.

adamjstewart commented:
> I didn't do fish (yet) but I'm sort of inclined to learn it b/c I played around with it and it was pretty awesome. 

I just switched this summer and I don't think I'll ever be able to switch back because fish is so much better. You can configure bash/zsh to be tolerable, but fish already has all of these features out-of-the-box.

> which I think is important if only because concretize is a very common command

I'm still in favor or merging spec/concretize/solve into a single environment-aware command. solve is the same as spec but with extra output (could be a spec flag) and concretize is the equivalent command for environments.
